das/ipfs_storage_service.go

```diff
@@ -0,0 +1,199 @@
+	"github.com/multiformats/go-multihash"
+	"github.com/offchainlabs/nitro/arbstate"
+	"github.com/offchainlabs/nitro/cmd/ipfshelper"
+	"github.com/offchainlabs/nitro/das/dastree"
+	"github.com/offchainlabs/nitro/util/pretty"
+	flag "github.com/spf13/pflag"
+)
+
+type IpfsStorageServiceConfig struct {
+	Enable      bool          `koanf:"enable"`
+	RepoDir     string        `koanf:"repo-dir"`
+	ReadTimeout time.Duration `koanf:"read-timeout"`
+	Profiles    string        `koanf:"profiles"`
+}
+
+var DefaultIpfsStorageServiceConfig = IpfsStorageServiceConfig{
+	Enable:      false,
+	RepoDir:     "",
+	ReadTimeout: time.Minute,
+	Profiles:    "test", // Default to test, see profiles here https://github.com/ipfs/kubo/blob/e550d9e4761ea394357c413c02ade142c0dea88c/config/profile.go
+}
+
+func IpfsStorageServiceConfigAddOptions(prefix string, f *flag.FlagSet) {
+	f.Bool(prefix+".enable", DefaultIpfsStorageServiceConfig.Enable, "enable storage/retrieval of sequencer batch data from IPFS")
+	f.String(prefix+".repo-dir", DefaultIpfsStorageServiceConfig.RepoDir, "directory to use to store the local IPFS repo")
+	f.Duration(prefix+".read-timeout", DefaultIpfsStorageServiceConfig.ReadTimeout, "timeout for IPFS reads, since by default it will wait forever. Treat timeout as not found")
+	f.String(prefix+".profiles", DefaultIpfsStorageServiceConfig.Profiles, "comma separated list of IPFS profiles to use")
+}
+
+type IpfsStorageService struct {
+	config     IpfsStorageServiceConfig
+	ipfsHelper *ipfshelper.IpfsHelper
+	ipfsApi    icore.CoreAPI
+}
+
+func NewIpfsStorageService(ctx context.Context, config IpfsStorageServiceConfig) (*IpfsStorageService, error) {
+	ipfsHelper, err := ipfshelper.CreateIpfsHelper(ctx, config.RepoDir, false, config.Profiles)
+	if err != nil {
+		return nil, err
+	}
+	return &IpfsStorageService{
+		config:     config,
+		ipfsHelper: ipfsHelper,
+		ipfsApi:    ipfsHelper.GetAPI(),
+	}, nil
+}
+
+func hashToCid(hash common.Hash) (cid.Cid, error) {
+	multiEncodedHashBytes, err := multihash.Encode(hash[:], multihash.KECCAK_256)
+	if err != nil {
+		return cid.Cid{}, err
+	}
+
+	_, multiHash, err := multihash.MHFromBytes(multiEncodedHashBytes)
+	if err != nil {
+		return cid.Cid{}, err
+	}
+
+	return cid.NewCidV1(cid.Raw, multiHash), nil
+}
+
+// GetByHash retrieves and reconstructs one batch's data, using IPFS to retrieve the preimages
+// for each chunk of data and the dastree nodes.
+func (s *IpfsStorageService) GetByHash(ctx context.Context, hash common.Hash) ([]byte, error) {
+	log.Trace("das.IpfsStorageService.GetByHash", "hash", pretty.PrettyHash(hash))
+
+	oracle := func(h common.Hash) ([]byte, error) {
+		thisCid, err := hashToCid(h)
+		if err != nil {
+			return nil, err
+		}
+
+		ipfsPath := path.IpfsPath(thisCid)
+		log.Trace("Retrieving IPFS path", "path", ipfsPath.String())
+
+		timeoutCtx, cancel := context.WithTimeout(ctx, s.config.ReadTimeout)
+		defer cancel()
+		rdr, err := s.ipfsApi.Block().Get(timeoutCtx, ipfsPath)
```

Comment:
I tried to figure out if there is a better way to work around hanging on GetByHash. There is an option of directly getting for a CID in Blockstore, without fetching from remotes, but I am not sure if that's what we want here.
If we didn't care about fetching the block from remotes then it could look like that:
```
block, err := s.ipfsHelper.GetNode().Blockstore.Get(timeoutCtx, thisCid)
...
data := block.RawData()
```

Reply:
That's cool to know. In this case we definitely want it to fetch from remote.